Src/DependencyCollector/Microsoft.AI.DependencyCollector/DependencyCollectorDiagnosticListener.cs

```diff
@@ -0,0 +1,190 @@
+        /// <seealso cref="IObserver{T}.OnNext(T)"/>
+        /// </summary>
+        /// <param name="value">The DiagnosticListener that exists when this listener was added to
+        /// the list, or a DiagnosticListener that got added after this listener was added.</param>
+        public void OnNext(DiagnosticListener value)
+        {
+            // Comes from https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/DiagnosticsHandlerLoggingStrings.cs#L12
+            if (value.Name == "HttpHandlerDiagnosticListener")
+            {
+                value.SubscribeWithAdapter(this);
+            }
+        }
+
+        /// <summary>
+        /// Notifies the observer that the provider has finished sending push-based notifications.
+        /// <seealso cref="IObserver{T}.OnCompleted()"/>
+        /// </summary>
+        public void OnCompleted()
+        {
+        }
+
+        /// <summary>
+        /// Notifies the observer that the provider has experienced an error condition.
+        /// <seealso cref="IObserver{T}.OnError(Exception)"/>
+        /// </summary>
+        /// <param name="error">An object that provides additional information about the error.</param>
+        public void OnError(Exception error)
+        {
+        }
+
+        /// <summary>
+        /// Get the DependencyTelemetry objects that are still waiting for a response from the dependency. This will most likely only be used for testing purposes.
+        /// </summary>
+        internal IEnumerable<DependencyTelemetry> PendingDependencyTelemetry
+        {
+            get { return pendingTelemetry.Values; }
+        }
+
+        /// <summary>
+        /// Diagnostic event handler method for 'System.Net.Http.Request' event.
+        /// </summary>
+        [DiagnosticName("System.Net.Http.Request")]
+        public void OnRequest(HttpRequestMessage request, Guid loggingRequestId)
+        {
+            if (request != null && request.RequestUri != null && !this.applicationInsightsUrlFilter.IsApplicationInsightsUrl(request.RequestUri.ToString()))
+            {
+                string httpMethod = request.Method.Method;
+                Uri requestUri = request.RequestUri;
+                string resourceName = requestUri.AbsolutePath;
+                if (!string.IsNullOrEmpty(httpMethod))
+                {
+                    resourceName = httpMethod + " " + resourceName;
+                }
+
+                DependencyTelemetry telemetry = new DependencyTelemetry();
+                this.client.Initialize(telemetry);
+                telemetry.Start();
+                telemetry.Name = resourceName;
+                telemetry.Target = requestUri.Host;
+                telemetry.Type = "Http";
+                telemetry.Data = requestUri.OriginalString;
+                this.pendingTelemetry.TryAdd(loggingRequestId, telemetry);
+
+                if (!request.Headers.Contains(RequestResponseHeaders.RequestContextSourceKey))
```

Comment:
can request.Headers ever be null?
Also this check is totally wrong. Use request.Headers.GetNameValueHeaderValue(RequestResponseHeaders.RequestContextHeader, RequestResponseHeaders.RequestContextSourceKey) == null. 

See .net implemenation.